source/non_matching/quadrature_generator.cc

```diff
@@ -0,0 +1,1364 @@
+        const Definiteness definiteness =
+          pointwise_definiteness(level_sets, center);
+
+        ExtendableQuadrature<dim> &quadrature =
+          this->q_partitioning.quadrature_by_definiteness(definiteness);
+
+        quadrature.push_back(center, box.volume());
+      }
+
+
+
+      template <int dim, int spacedim>
+      void
+      QGenerator<dim, spacedim>::set_1D_quadrature(const unsigned int q_index)
+      {
+        AssertIndexRange(q_index, this->q_collection1D->size());
+
+        this->q_index = q_index;
+        low_dim_algorithm.set_1D_quadrature(q_index);
+        up_through_dimension_creator.set_1D_quadrature(q_index);
+      }
+
+
+
+      template <int spacedim>
+      QGenerator<1, spacedim>::QGenerator(
+        const hp::QCollection<1> &      q_collection1D,
+        const AdditionalQGeneratorData &additional_data)
+        : QGeneratorBase<1, spacedim>(q_collection1D, additional_data)
+        , root_finder(
+            RootFinder::AdditionalData(additional_data.root_finder_tolerance,
+                                       additional_data.max_root_finder_splits))
+      {
+        Assert(q_collection1D.size() > 0,
+               ExcMessage("Incoming quadrature collection is empty."));
+      }
+
+
+
+      template <int spacedim>
+      void
+      QGenerator<1, spacedim>::create_quadrature(
+        const std::vector<std::reference_wrapper<const Function<1>>>
+          &                   level_sets,
+        const BoundingBox<1> &box,
+        const unsigned int    n_box_splits)
+      {
+        (void)n_box_splits;
+
+        roots.clear();
+        root_finder.find_roots(level_sets, box, roots);
+
+        const Quadrature<1> &quadrature1D =
+          (*this->q_collection1D)[this->q_index];
+
+        distribute_points_between_roots(quadrature1D,
+                                        box,
+                                        roots,
+                                        zero_dim_point,
+                                        unit_weight,
+                                        direction,
+                                        level_sets,
+                                        this->additional_data,
+                                        this->q_partitioning);
+
+        if (spacedim == 1)
+          this->create_surface_points(level_sets);
+      }
+
+
+
+      template <int spacedim>
+      void
+      QGenerator<1, spacedim>::create_surface_points(
+        const std::vector<std::reference_wrapper<const Function<1>>>
+          &level_sets)
+      {
+        Assert(level_sets.size() == 1, ExcInternalError());
+
+        for (unsigned int i = 0; i < roots.size(); ++i)
+          {
+            // A surface integral in 1D is just a point evaluation,
+            // so the weight is always 1.
+            const double   weight = 1;
+            const Point<1> point(roots[i]);
+
+            Tensor<1, 1> normal        = level_sets[0].get().gradient(point);
```

Comment:
Can you call this `gradient` or something like that? It should be close to a normal vector for a good signed distance function, but as the assert shows we cannot guarantee it here. As a side note, can you think of a case where the hardcoded tolerance `1e-11` below would be bad and we would need a scaling? We often try to make comparisons in the library also work when say the domain has extent 1e20 (or 1e-20). But I guess for a signed distance function we would be really eliminate the h scale, so I think we are fine.

Reply:
I can change the name, but I normalize the tensor just a few lines below. So are you sure that's what you want?

Regarding the assert, I guess you can always come up with a function where it fails:

\psi(x) = k (x- x_{root}),        k --> 0

The main reason for the assert here is to give a better error message in the case that we try to normalize something that gives a NaN. Given that, I'm not sure what a good value in the assert would be. Do you think it would be better to avoid the division completely with something like this?
```c++
normal[0] = boost::math::sign(gradient[0])
```

Reply:
I agree with your point - let's just keep these things as they are now. If we ever hit a case with a bad error (and signed distance functions should by construction be scale-invariant), we might search for ideas then.